src/delphi/train/training.py

```diff
@@ -0,0 +1,276 @@
+
+from datetime import datetime
+from dataclasses import dataclass
+import os 
+import torch
+import time
+from contextlib import nullcontext
+import math
+@dataclass
+def TrainingConfig(config):
+    # -----------------------------------------------------------------------------
+    # I/O
+    out_dir: str = "out"
+    eval_interval: int = 2000
+    log_interval:int = 1
+    eval_iters:int = 100
+    eval_only:bool = False  # if True, script exits right after the first eval
+    always_save_checkpoint:bool = False  # if True, always save a checkpoint after each eval
+    init_from:bool = "scratch"  # 'scratch' or 'resume'
+    # wandb logging
+    wandb_log:bool = False  # disabled by default
+    wandb_project:str = "llamac"
+    wandb_run_name:str = "run" + datetime.now().strftime("%Y_%m_%d_%H_%M_%S")
+    # data
+    batch_size:int = 128  # if gradient_accumulation_steps > 1, this is the micro-batch size
+    max_seq_len:int = 256
+    vocab_source:str = "llama2" # llama2|custom; use Lllama 2 vocab from Meta, or custom trained
+    vocab_size:str = 32000 # the Llama 2 tokenizer has 32K tokens
+    # model
+    dim:int = 288
+    n_layers:int = 6
+    n_heads:int = 6
+    n_kv_heads:int = 6
+    multiple_of:int = 32
+    dropout:int = 0.0
+    # adamw optimizer
+    gradient_accumulation_steps:int = 4  # used to simulate larger batch sizes
+    learning_rate:float = 5e-4  # max learning rate
+    max_iters:int = 100000  # total number of training iterations
+    weight_decay:float = 1e-1
+    beta1:float = 0.9
+    beta2:float = 0.95
+    grad_clip:float = 1.0  # clip gradients at this value, or disable if == 0.0
+    # learning rate decay settings
+    decay_lr:bool = True  # whether to decay the learning rate
+    warmup_iters:int = 1000  # how many steps to warm up for
+    # system
+    device:str = "cuda"  # examples: 'cpu', 'cuda', 'cuda:0', 'cuda:1' etc., or try 'mps' on macbooks
+    dtype:str = "bfloat16"  # float32|bfloat16|float16
+    compile:bool = True  # use PyTorch 2.0 to compile the model to be faster
+    # -----------------------------------------------------------------------------
+    config_keys = [
+        k
+        for k, v in globals().items()
+        if not k.startswith("_") and isinstance(v, (int, float, bool, str))
+    ]
+    exec(open("configurator.py").read())  # overrides from command line or config file
+    config = {k: globals()[k] for k in config_keys}  # will be useful for logging
```

Comment:
I'm confused by what's going on here, I might just be ignorant of some stuff or maybe the code is in the middle of editing?

- there's a @dataclass decorator on a function
- the function is named in class-style CamelCase
- the `config` argument isn't used and instead appears to be overwritten by assignment
- there's some magic going on with a `configurator.py` file that's not in the repo
- we're loading globals into config_keys, but there don't appear to be any meaningful globals in this script (is it being called from somewhere else?)

Reply:
The original llama2.c just had each of them as a global variable that gets overwritten from command line arguments using the "configurator.py" file. Karpathy doesn't like writing config.X before each training attribute, wherefore he set up this workaround, but @SrGonao wrapped it into a data class. I am fine with both settings but we should decide on either of them. @jettjaniak - what do you prefer?

Reply:
probably
- define a TrainingConfig dataclass (not function with a dataclass decorator) in train/training_config.py
- make sure it's type-checked (maybe `beartype_this_package()` in `src/delphi/__init__.py` is already doing it?), that should work:
```
@beartype
@dataclass
class A:
```

but then we want a way to initialize this class from CLI args via argparse or from a config file, so I'm not sure what would be the best way - maybe there is a dedicated solution / package for this - ask some AI

Reply:
we definitely want TrainingConfig object that we can pass to functions, that's type-checked and can be easily initialized from CLI or from file w/o having to define all the arguments 2 or 3 times in code

Reply:
This looks like a pretty good solution for the argument parsing, h/t gpt4:

```python3.10
from dataclasses import fields
import argparse

parser = argparse.ArgumentParser()

for field in fields(TrainingConfig):
    parser.add_argument(f"--{field.name}", type=field.type)

training_config_args = parser.parse_args()

my_config = TrainingConfig(**vars(training_config_args))
```

Just a little tweaking to get config-from-file support. Have to run, back later

Reply:
[Hydra](https://hydra.cc/docs/intro/) could be useful. It offers config file and command line arguments, and would make our configs much smaller using inheritance

Reply:
#48